### PR TITLE
No U Turn Sampler

### PR DESCRIPTION
**Pull request by jsalvatier**
Implemented the [No U Turn Sampler](http://arxiv.org/pdf/1111.4246v1.pdf) as [`NoUTurn`](https://github.com/pymc-devs/pymc/blob/nuts2/pymc/step_methods/nuts.py). I did a very straightforward implementation of Algorithm 6 - Efficient No-U-Turn Sampler with Dual Averaging.

It seems to be working well on the stochastic volatility model (not surprising since its one of the examples in the paper). It even reduces the sampling time by about half.

Since NUTS seems to work pretty well for Stan, should this be the new default and convert the examples using HMC to NUTS?

Still needed: 
- [x] Tests
- [x] Docs

### CODE REVIEW

File: pymc/step_methods/hmc.py [hunk 1]
```diff
@@ -10,7 +10,9 @@
 import numpy as np
 from scipy.sparse import issparse
 
-__all__ = ['HamiltonianMC']
+from collections import namedtuple
+
+__all__ = ['HamiltonianMC', 'leapfrog','energy', 'Hamiltonian', 'bern']
```
Comment (twiecki) [inline pymc/step_methods/hmc.py]: Should all of these be "public"?

File: pymc/step_methods/hmc.py [hunk 2]
```diff
@@ -46,7 +48,7 @@ def __init__(self, vars, C, step_scale=.25, path_length=2., is_cov=False, step_r
 
         self.step_size = step_scale / n ** (1 / 4.)
 
-        self.potential = quad_potential(C, is_cov)
+        self.potential = quad_potential(C, is_cov, as_cov = False)
```
Comment (twiecki) [inline pymc/step_methods/hmc.py:51]: Pedantic: pep8 says these should be `as_cov=False` (without the spaces).

File: pymc/step_methods/quadpotential.py [hunk 3]
```diff
@@ -9,24 +9,21 @@
 __all__ = ['quad_potential', 'ElemWiseQuadPotential', 'QuadPotential',
            'QuadPotential_Inv', 'isquadpotential']
 
+def quad_potential(C, is_cov, as_cov):
```
Comment (twiecki) [inline pymc/step_methods/quadpotential.py:12]: the last two seem a little confusing. Maybe some comments/doc-string?

File: pymc/step_methods/hmc.py (hunk 1 above)
Comment (jsalvatier) [inline pymc/step_methods/hmc.py]: Maybe not. Nuts uses them, but we can just pull them directly.

**Comment by twiecki** [comment]
Looks good! I also tried it with the GLM and it indeed seems to be more stable than non-adaptive HMC.

**Comment by fonnesbeck** [comment]
A minor enhancement suggestion: should we call it `NUTS` instead of `NoUTurn`? Simply a convenience, and consistent with what the Stan folks call it.

**Comment by twiecki** [comment]
I agree.

**Comment by jsalvatier** [comment]
Sounds good. You want to make the change Thomas?

**Comment by twiecki** [comment]
Sure.

**Comment by twiecki** [comment]
Committed via ea90dfac60aa128e201312973eddd6a1134b089c

**Pull request merged**